[quant][graphmode] Move numerics changing passes before finalize

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37515 [quant][graphmode] Support a new category of ops in graph mode quantization
* #37555 [quant][graphmode][refactor] rename SwapDequant and refactor code handling general ops
* **#37514 [quant][graphmode] Move numerics changing passes before finalize**

Summary:
This is to constrain all numerics changing operations in insert quant dequant pass

Test Plan:
python test/test_quantization.py TestQuantizeScriptJitPasses

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D21364008](https://our.internmc.facebook.com/intern/diff/D21364008)

jerryzh168 committed:
[quant][graphmode] Move numerics changing passes before finalize

Summary:
This is to constrain all numerics changing operations in insert quant dequant pass

Test Plan:
python test/test_quantization.py TestQuantizeScriptJitPasses

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Move numerics changing passes before finalize"


Summary:
This is to constrain all numerics changing operations in insert quant dequant pass

Test Plan:
python test/test_quantization.py TestQuantizeScriptJitPasses

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Move numerics changing passes before finalize"


Summary:
This is to constrain all numerics changing operations in insert quant dequant pass

Test Plan:
python test/test_quantization.py TestQuantizeScriptJitPasses

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Move numerics changing passes before finalize"


Summary:
This is to constrain all numerics changing operations in insert quant dequant pass

Test Plan:
python test/test_quantization.py TestQuantizeScriptJitPasses

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Move numerics changing passes before finalize"


Summary:
This is to constrain all numerics changing operations in insert quant dequant pass

Test Plan:
python test/test_quantization.py TestQuantizeScriptJitPasses

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on test/quantization/test_quantize_script.py:
```diff
@@ -1133,7 +1107,6 @@ def forward(self, x):
         model = torch.jit.script(M()).eval()
         model = quantize_script(model, qconfig_dict, _test_only_eval_fn, [data], inplace=False, debug=True)
         FileCheck().check_not("quantized::conv2d") \
-                   .check("CallMethod") \
```
Why do we not check for CallMethod anymore?

jerryzh168 replied:
yeah, we'll inline the graph in convert so CallMethod from sub module is gone

supriyar commented on test/quantization/test_quantize_script.py:
```diff
@@ -724,32 +724,16 @@ def forward(self, x):
-                       .check_not(quant_func) \
-                       .check("return") \
-                       .run(str(get_module_method(m, 'conv', '_conv_forward').graph))
+            FileCheck().check_count(quant_func, 3, exactly=True) \
```
I think dynamic quant tests need to be updated too

supriyar commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -361,11 +361,12 @@ void export_opnames(const script::Module& m, std::set<std::string>& opnames) {
-      row->elements().at(0).toStringRef());
+        row->elements().at(0).toStringRef() == "operators",
+        "Expected operators but found ",
+        row->elements().at(0).toStringRef());
```
Should this change be here?

jerryzh168 replied:
this is clang format changes, got this when I run clang-format-all.py

jerryzh168 committed:
Update on "[quant][graphmode] Move numerics changing passes before finalize"


Summary:
This is to constrain all numerics changing operations in insert quant dequant pass

Test Plan:
python test/test_quantization.py TestQuantizeScriptJitPasses

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Move numerics changing passes before finalize"


Summary:
This is to constrain all numerics changing operations in insert quant dequant pass

Test Plan:
python test/test_quantization.py TestQuantizeScriptJitPasses

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1818,6 +1820,108 @@ void InsertQuantDeQuantHelper::quantizeTensors(
+}
+
+void InsertQuantDeQuantHelper::quantizeGeneralOps(Module& module) {
+  SwapFunctionalLinear(module);
```
Shouldnt this step be done prior to observer insertion?

jerryzh168 replied:
this is to prevent inlining the linear call functions, we don't need to do this swap before inserting observers

raghuramank100 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1682,6 +1682,8 @@ class InsertQuantDeQuantHelper {
     is_dynamic_ = is_dynamic;
   }
 
+  void quantizeGeneralOps(Module& module);
```
Can we have a more descriptive name for this function:  Can we call this propagateQuantDeQuant?

jerryzh168 replied:
Sure

jerryzh168 replied:
you can find the comment in quantization.h

jerryzh168 replied:
renaming is done in next PR

jerryzh168 committed:
Update on "[quant][graphmode] Move numerics changing passes before finalize"


Summary:
This is to constrain all numerics changing operations in insert quant dequant pass

Test Plan:
python test/test_quantization.py TestQuantizeScriptJitPasses

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Move numerics changing passes before finalize"


Summary:
This is to constrain all numerics changing operations in insert quant dequant pass

Test Plan:
python test/test_quantization.py TestQuantizeScriptJitPasses

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
